File: src/Vpiska.ApiChat/Connectors/ChatConnector.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using Serilog;
using Vpiska.WebSocket;

namespace Vpiska.ApiChat.Connectors
{
    public sealed class ChatConnector : IWebSocketConnector
    {
        private const string ParamName = "eventId";

        private readonly ILogger _logger;
        private readonly UserConnectionsStorage _storage;

        public ChatConnector(ILogger logger, UserConnectionsStorage storage)
        {
            _logger = logger;
            _storage = storage;
        }
        
        public void OnConnect(Guid connectionId, WebSocketUserContext userContext, Dictionary<string, string> queryParams)
        {
            if (Guid.TryParse(queryParams[ParamName], out var eventId))
            {
                if (!_storage.IsEventExist(eventId))
                {
                    if (_storage.TryCreateUserGroup(eventId))
                    {
                        //todo subscribe to event
                    }
                }

                if (!_storage.TryAddUserContext(eventId, connectionId, userContext))
                {
                    _logger.Warning("can't add user context. EventId: {eventId}. ConnectionId: {connectionId}", eventId,
                        connectionId);
                }
            }
        }

        public void OnDisconnect(Guid connectionId, Dictionary<string, string> queryParams)
        {
            if (Guid.TryParse(queryParams[ParamName], out var eventId))
            {
                if (_storage.TryRemoveUserContext(eventId, connectionId))
                {
                    if (_storage.GetUsersCount(eventId) == 0)
                    {
                        if (_storage.TryRemoveUserGroup(eventId))
                        {
                            //todo unsubscribe from event
                        }
                    }
                }
            }
        }
    }

    public sealed class UserConnectionsStorage
    {
        private readonly ConcurrentDictionary<Guid, ConcurrentDictionary<Guid, WebSocketUserContext>>
            _eventConnections;
        private readonly ConcurrentDictionary<Guid, Guid> _userConnections;

        public UserConnectionsStorage()
        {
            _eventConnections = new ConcurrentDictionary<Guid, ConcurrentDictionary<Guid, WebSocketUserContext>>();
            _userConnections = new ConcurrentDictionary<Guid, Guid>();
        }

        public bool IsEventExist(Guid eventId) => _eventConnections.ContainsKey(eventId);

        public int GetUsersCount(Guid eventId) =>
            _eventConnections.TryGetValue(eventId, out var users) ? users.Count : 0;

        public bool TryCreateUserGroup(Guid eventId) =>
            _eventConnections.TryAdd(eventId, new ConcurrentDictionary<Guid, WebSocketUserContext>());

        public bool TryRemoveUserGroup(Guid eventId) =>
            _eventConnections.TryRemove(eventId, out _);

        public bool TryAddUserContext(Guid eventId, Guid connectionId, WebSocketUserContext context) =>
            _eventConnections.TryGetValue(eventId, out var users)
            && users.TryAdd(connectionId, context)
            && _userConnections.TryAdd(context.UserId, connectionId);

        public bool TryRemoveUserContext(Guid eventId, Guid connectionId) =>
            _eventConnections.TryGetValue(eventId, out var users)
            && users.TryRemove(connectionId, out var user)
            && _userConnections.TryRemove(user.UserId, out _);
    }
}

File: src/Vpiska.Application/Event/WebSocket/ChatConnector.fs
namespace Vpiska.Application.Event.WebSocket

open System
open System.Threading.Tasks
open Serilog
open FSharp.Control.Tasks
open Vpiska.Application.Event
open Vpiska.Application.Event.CommandHandler
open Vpiska.Domain.Event
open Vpiska.Infrastructure.Websocket

type ChatConnector(logger: ILogger, storage: UserConnectionsStorage, persistence: EventPersistence) =
    
    let queryParamName = "eventId"
    
    let handle = CommandHandler.handle persistence
    
    let mapErrors errors = errors |> Array.map Errors.mapAppError |> Array.fold (fun acc error -> $"{error}. {acc}") ""
    
    let connectUser (eventId: string) (connectionId: Guid) (userContext: WebSocketUserContext) =
        task {
            match storage.TryAddUserContext (eventId, connectionId, userContext) with
            | false -> logger.Warning("can't add user context. EventId: {eventId}. ConnectionId: {connectionId}", eventId, connectionId)
            | true ->
                let args = { EventId = eventId; UserId = userContext.UserId; Name = userContext.Username; ImageId = userContext.UserImageId }
                match! args |> Command.LogUserInChat |> handle with
                | Error errors ->
                    let message = mapErrors errors
                    logger.Warning("can't connect user to event. Message: {message}", message)
                | Ok _ -> return ()
        } :> Task
        
    let disconnectUser (eventId: string) (userId: string) =
        task {
            let command = { EventId = eventId; UserId = userId } |> Command.LogoutUserFromChat
            match! handle command with
            | Error errors ->
                let message = mapErrors errors
                logger.Warning("can't disconnect user from event. Message: {message}", message)
            | Ok _ -> return ()
        } :> Task
    
    interface IWebSocketConnector with
        member _.OnConnect(connectionId, userContext, queryParams) =
            task {
                let eventId = queryParams.[queryParamName]
                match storage.Exist eventId with
                | true -> do! connectUser eventId connectionId userContext
                | false ->
                    match storage.TryCreateUserGroup eventId with
                    | false -> logger.Warning("can't create user group. EventId: {eventId}. ConnectionId: {connectionId}", eventId, connectionId)
                    | true ->
                        let command = { EventId = eventId } |> Command.Subscribe
                        match! handle command with
                        | Error errors ->
                            let message = mapErrors errors
                            logger.Warning("can't subscribe for event. Message: {message}", message)
                        | Ok _ -> do! connectUser eventId connectionId userContext
            } :> Task
            
        
        member _.OnDisconnect(connectionId, queryParams) =
            task {
                let eventId = queryParams.[queryParamName]
                let mutable userContext = null
                match storage.TryRemoveUserContext (eventId, connectionId, &userContext) with
                | false -> logger.Warning("can't remove user context. EventId: {eventId}. ConnectionId: {connectionId}", eventId, connectionId)
                | true ->
                    do! disconnectUser eventId userContext.UserId
                    match storage.GetUsersCount eventId with
                    | 0 ->
                        match storage.TryRemoveUserGroup eventId with
                        | false -> logger.Warning("can't remove user group. EventId: {eventId}. ConnectionId: {connectionId}", eventId, connectionId)
                        | true ->
                            let command = { EventId = eventId } |> Command.Unsubscribe
                            match! handle command with
                            | Error errors ->
                                let message = mapErrors errors
                                logger.Warning("can't unsubscribe from event. Message: {message}", message)
                            | Ok _ -> return ()
                    | _ -> return ()
            } :> Task
